test/unit/eigensolver/test_eigensolver.cpp

```diff
@@ -128,18 +129,37 @@ void testEigensolver(const blas::Uplo uplo, const SizeType m, const SizeType mb)
   Matrix<T, Device::CPU> mat_a_h(reference.distribution());
   copy(reference, mat_a_h);
 
-  eigensolver::EigensolverResult<T, D> ret = [&]() {
-    MatrixMirror<T, D, Device::CPU> mat_a(mat_a_h);
-    return eigensolver::eigensolver<B>(uplo, mat_a.get());
-  }();
+  if constexpr (!preallocated) {
+    eigensolver::EigensolverResult<T, D> ret = [&]() {
+      MatrixMirror<T, D, Device::CPU> mat_a(mat_a_h);
+      return eigensolver::eigensolver<B>(uplo, mat_a.get());
+    }();
```

Comment:
You can avoid some code duplication if you move the constexpr branch inside the lambda
```suggestion
  eigensolver::EigensolverResult<T, D> ret = [&]() {
    MatrixMirror<T, D, Device::CPU> mat_a(mat_a_h);
    if constexpr (preallocated) {
      const SizeType size = mat_a_h.size().rows();
      Matrix<BaseType<T>, D> eigenvalues(LocalElementSize(size, 1),
                                     TileElementSize(mat_a_h.blockSize().rows(), 1));
      Matrix<T, D> eigenvectors(LocalElementSize(size, size), mat_a_h.blockSize());
      eigensolver::eigensolver<B>(uplo, mat_a.get(), eigenvectors, eigenvalues);
      return eigensolver::EigensolverResult<T, D>(std::move(eigenvalues), std::move(eigenvectors));
    else {
      return eigensolver::eigensolver<B>(uplo, mat_a.get());
    }
  }();
```